driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedTest.java

```diff
@@ -449,6 +502,186 @@ protected boolean terminateLoop() {
+                        entities.getServerListener(clientId));
+                break;
+            case "poolClearedEvent":
+            case "poolReadyEvent":
+                context.getEventMatcher().waitForConnectionPoolEvents(clientId, event, count, entities.getConnectionPoolListener(clientId));
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported event: " + eventName);
+        }
+
+        return OperationResult.NONE;
+    }
+
+    private OperationResult executeAssertEventCount(final UnifiedTestContext context, final BsonDocument operation) {
+        BsonDocument arguments = operation.getDocument("arguments");
+        String clientId = arguments.getString("client").getValue();
+        BsonDocument event = arguments.getDocument("event");
+        String eventName = event.getFirstKey();
+        int count = arguments.getNumber("count").intValue();
+
+        switch (eventName) {
+            case "serverDescriptionChangedEvent":
+                context.getEventMatcher().assertServerDescriptionChangeEventCount(clientId, event, count,
+                        entities.getServerListener(clientId).getServerDescriptionChangedEvents());
+                break;
+            case "poolClearedEvent":
+            case "poolReadyEvent":
+                context.getEventMatcher().assertConnectionPoolEventCount(clientId, event, count,
+                        entities.getConnectionPoolListener(clientId).getEvents());
+                break;
+            default:
+                throw new UnsupportedOperationException("Unsupported event: " + eventName);
+        }
+
+        return OperationResult.NONE;
+    }
+
+    private OperationResult executeWaitPrimaryChange(final UnifiedTestContext context, final BsonDocument operation) {
+        context.getAssertionContext().push(ContextElement.ofWaitForPrimaryChange());
+        BsonDocument arguments = operation.getDocument("arguments");
+        MongoClient client = entities.getClient(arguments.getString("client").getValue());
+        ClusterDescription priorClusterDescription =
+                entities.getTopologyDescription(arguments.getString("priorTopologyDescription").getValue());
+        ClusterDescription currentClusterDescription = client.getClusterDescription();
+        long timeoutNanos =
+                TimeUnit.NANOSECONDS.convert(arguments.getNumber("timeoutMS", new BsonInt32(10000)).longValue(), TimeUnit.MILLISECONDS);
+        long startTime = System.nanoTime();
+        while (primaryIsSame(priorClusterDescription, currentClusterDescription) || noPrimary(currentClusterDescription)) {
+            if (System.nanoTime() - startTime > timeoutNanos) {
+                fail(context.getAssertionContext().getMessage("Timed out waiting for primary change"));
+            }
+            try {
+                //noinspection BusyWait
```

Comment:
We could avoid busy waiting with a more complicated implementation of ServerListener, but it didn't seem worth the effort, so just looping instead with a busy wait.

Reply:
Ack - sounds good.